### PR TITLE
Fixes for compilation without native dynlink

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix

I fell on problems when compiling without native dynlink and this PR fixes them:
- `DECLARE PLUGIN` missing for `ltac2_plugin` (note that this impacts also 8.11)
- `DECLARE PLUGIN` missing for `ssrsearch_plugin`
- making fetching digests aware of the possibility of no native dynlink

Incidentally, @coq/ci-maintainers, how easy would it be to add a regression test for configuration with `-natdynlink no`?

- [ ] Entry added in the changelog (not needed?)

### CODE REVIEW

**Comment by ejgallego** [comment]
> Incidentally, @coq/ci-maintainers, how easy would it be to add a regression test for configuration with `-natdynlink no`?

I am not sure it is worth supporting `natdynlink no` at all; IIANM the Coq produced there is of limited utility as it cannot load any plugins anyway [other than the statically linked ones], IMHO, such arquitectures should just run the bytecode version.

**Comment by herbelin** [comment]
It is still useful for debugging for which I actually directly modify the Makefile. Doing so, it cannot load Ltac2 because of the missing `DECLARE PLUGIN`, so these `DECLARE PLUGIN` are needed at least for static inclusion of the byte version of plugins (and thus of debugging).

**Comment by herbelin** [comment]
How is static linking of bytecode plugins supported with dune for debugging?

**Comment by ejgallego** [comment]
> It is still useful for debugging for which I actually directly modify the Makefile. Doing so, it cannot load Ltac2 because of the missing `DECLARE PLUGIN`, so these `DECLARE PLUGIN` are needed at least for static inclusion of the byte version of plugins (and thus of debugging).

Shouldn't the static linking be an option separate of natdynlink?

> How is static linking of bytecode plugins supported with dune for debugging?

I understand ocamldbg works with plugins in recent OCaml releases [not 100% sure about this tho] , you can just pass the plugins to link statically in the `(exectuable...)` stanza in topbin, if that's a common thing we can have configure generate a list of things to link statically and have the dune file `(include ...)` that list.

**Comment by herbelin** [comment]
> I understand ocamldbg works with plugins in recent OCaml releases [not 100% sure about this tho]

It is [apparently](https://github.com/ocaml/ocaml/issues/6792) in 4.10.

> Shouldn't the static linking be an option separate of natdynlink?

Yes, makes sense. So, you mean adding a configure option `-static-plugins` or something like that?

**Comment by ejgallego** [comment]
> Yes, makes sense. So, you mean adding a configure option `-static-plugins` or something like that?

Indeed, we could do that, I think the Makefile system already has that; for Dune should be easy. On the other hand if newer OCaml do solve that maybe it is not so prioritary. Anyways this PR is fairly orthogonal, so let's go ahead with it.

File: vernac/mltop.ml [hunk 1]
```diff
@@ -309,6 +309,9 @@ type ml_module_object = {
 }
 
 let add_module_digest m =
+  if not has_dynlink then
+    m, NoDigest
+  else
```
Comment (ejgallego) [inline vernac/mltop.ml:314]: Out of curiosity @herbelin, why is this needed?

Comment (herbelin) [reply vernac/mltop.ml:314]: Because otherwise `file_of_name` fails with a user error at `Declare ML Module` time if there is effectively no plugin.

Another solution would be to add `UserError` in the `try with`.

Comment (ejgallego) [reply vernac/mltop.ml:314]: I see, thanks for the explanation; I was planning to do a pass in mltop [in particular to add support for dependency loading of plugins, a long requested feature]

I will keep this comment in mind when doing the pass; of course if you are inclined to add the comment on the code itself that would be nice.

**Review by ejgallego — APPROVED** [review]
Trivial, LGTM.

**Review by ppedrot — APPROVED** [review]
